packages/core/src/components/widgets/AltDateWidget.tsx

```diff
@@ -17,23 +16,47 @@ function rangeOptions(start: number, stop: number) {
   return options;
 }
 
-function readyForChange(state: any) {
-  return Object.keys(state).every((key) => state[key] !== -1);
+function readyForChange(state: DateObject) {
+  return Object.values(state).every((value) => value !== -1);
 }
 
-function DateElement(props: any) {
-  const {
-    type,
-    range,
-    value,
-    select,
-    rootId,
-    disabled,
-    readonly,
-    autofocus,
-    registry,
-    onBlur,
-  } = props;
+function dateElementProps(
+  state: DateObject,
+  time: boolean,
+  yearsRange: [number, number] = [1900, new Date().getFullYear() + 2]
+) {
+  const { year, month, day, hour, minute, second } = state;
+  const data = [
+    {
+      type: "year",
+      range: yearsRange,
+      value: year,
+    },
+    { type: "month", range: [1, 12], value: month },
+    { type: "day", range: [1, 31], value: day },
+  ] as { type: string; range: [number, number]; value: number | undefined }[];
+  if (time) {
+    data.push(
+      { type: "hour", range: [0, 23], value: hour },
+      { type: "minute", range: [0, 59], value: minute },
+      { type: "second", range: [0, 59], value: second }
+    );
+  }
+  return data;
+}
+
+function DateElement({
```

Comment:
Jimmy, here are the raw diffs for the changes I made to avoid the `any` type:
```
-function DateElement(props: any) {
+type DateElementProps<T, F> =
+  Pick<WidgetProps<T, F>, "value" | "disabled" | "readonly" | "autofocus" | "registry" | "onBlur" | "onFocus"> & {
+  rootId: string;
+  select: (property: keyof DateObject, value: any) => void;
+  type: string;
+  range: [number, number];
+}
+
+function DateElement<T, F>(props: DateElementProps<T, F>) {
   const {
     type,
     range,
@@ -33,6 +40,7 @@ function DateElement(props: any) {
     autofocus,
     registry,
     onBlur,
+    onFocus,
   } = props;
   const id = rootId + "_" + type;
   const { SelectWidget } = registry.widgets;
@@ -47,8 +55,11 @@ function DateElement(props: any) {
       disabled={disabled}
       readonly={readonly}
       autofocus={autofocus}
-      onChange={(value: any) => select(type, value)}
+      onChange={(value: any) => select(type as keyof DateObject, value)}
       onBlur={onBlur}
+      onFocus={onFocus}
+      registry={registry}
+      label=""
     />
   );
 }
@@ -57,7 +68,7 @@ function DateElement(props: any) {
 class AltDateWidget<T = any, F = any> extends Component<
   WidgetProps<T, F>,
   DateObject
-> {
+  > {
   static defaultProps = {
     time: false,
     disabled: false,
@@ -98,7 +109,7 @@ class AltDateWidget<T = any, F = any> extends Component<
       { [property]: typeof value === "undefined" ? -1 : value } as Pick<
         DateObject,
         keyof DateObject
-      >,
+        >,
       () => {
         // Only propagate to parent state if we have a complete date{time}
         if (readyForChange(this.state)) {
@@ -150,13 +161,13 @@ class AltDateWidget<T = any, F = any> extends Component<
   }
 
   render() {
-    const { id, disabled, readonly, autofocus, registry, onBlur, options } =
+    const { id, disabled, readonly, autofocus, registry, onBlur, onFocus, options } =
       this.props;
     return (
       <ul className="list-inline">
         {this.dateElementProps.map((elemProps, i) => (
           <li key={i}>
-            <DateElement
+            <DateElement<T, F>
               rootId={id}
               select={this.onChange}
               {...elemProps}
@@ -164,6 +175,7 @@ class AltDateWidget<T = any, F = any> extends Component<
               readonly={readonly}
               registry={registry}
               onBlur={onBlur}
+              onFocus={onFocus}
               autofocus={autofocus && i === 0}
             />
           </li>
```

Reply:
applied!